tutorial_test.go

```diff
@@ -32,12 +33,30 @@ type Person struct {
 	LastName  string `edgedb:"last_name"`
 }
 
+type OptionalYear struct {
+	year    int64
+	missing bool
+}
```

Comment:
I don't like this API :(

Perhaps we should take a similar route to https://github.com/markphelps/optional -- define an "optional" struct for all primitive types that EdgeDB can return with a `Get()` method defined on those structs.

So instead of defining your own unmarshaller and struct every time you see an optional scalar you could use a predefined struct, e.g. `edgedb.OptionalInt64`.

Reply:
We can do that in addition to what is here as another level of abstraction, but I think we need to expose something like this also because 1) shapes can be optional and 2) we support user defined scalar codecs.

Reply:
Yeah, I'm not suggesting that we should hide the API, just make it more convenient to work with.

Reply:
Question: can we declare our own `Optional` struct and let people embed it into optional structs? Like this (in psudo-code):

```go
// edgedb-go:

type Optional struct {
    __loaded__ bool
}

func (l *Optional) isEmpty() bool {
   return l.__loaded__
}
```

so that users would be able to do this:

```
type Movie struct {
   edgedb.Optional
   ....
}

// and later do

if movie.isEmpty() { ... }
```

?

Here's a runnable example: https://play.golang.org/p/mXK2jNDzhLQ

Reply:
Yes we can. Good idea!

Reply:
Then with this and with the scalar optional wrappers dealing with optionals should become easier

Reply:
Hm, I think common pattern in go is to use pointers? I.e. if you need optional int, you use `field *int`. And then you check whether `x.field != nil` for the presence of the value.

Reply:
That is a pattern, but I'm not sure it is THE pattern. The stdlib uses structs like this for the same purpose in [database/sql](https://golang.org/pkg/database/sql/#NullBool). When using pointers it is easy to forget to check if the pointer is `nul`. This API forces users to think more about what they are doing and hopefully makes it more obvious to readers what is going on.

Reply:
The problem with scalar pointers in structs is that you cannot take an address of an expression, so working with structs becomes a nightmare, e.g.:

```go
bool_field := some_bool_expr
struct = Struct{bool_field: &bool_field}
```

instead of

```
Struct{bool_field: some_bool_expr}
```

This obviously becomes a lot less pleasant as the number of fields grows.  Another pitfall is copying, as a copy of a Struct will reuse pointers to values instead of copying the values, so you need to write explicit copy functions.

Reply:
@elprans 
1. This is mostly for rows returned from the database, right? So this ergonomic doesn't play role. For input: you rarely use optional variables in queries, and if you do, they go from somewhere, not constant right right in `.query()` call (you would put the constant in the query itself, probably, or just don't do it optional, since you know the value).
2. With copying I think there is the same deep vs shallow copy for arrays, right? So generally this is a known quirk in go. On the other hand, I don't think "fetch then clone then edit structure" is a common pattern. Even just editing a structure in place is not common, unless we encourage using EdgeDB in ORM style. And even when using ORM cloning is somewhat rarely used in my practice (part of that is confusing identity, which has to be rewritten during the clone anyway, so special clone function anyway).

Note that not forgetting to fill `__loaded__` is generally a more real problem if creating such structures would be a common code. And not forgetting to check `__loaded__` can be a bigger problem (i.e. a security issue) than occasionally crashing on `nil` (there should be a go lint to check the latter, no?)

Reply:
Imo we don't need the pointers based api and my suggestion about adding optional versions of the primitive types + and optional struct marker should suffice.